packages/@glimmer/validator/lib/debug.ts

```diff
@@ -121,13 +121,18 @@ if (DEBUG) {
    *
    * TODO: Only throw an error if the `track` is consumed.
    */
-  runInTrackingTransaction = (fn: () => void, debugLabel?: string | false) => {
+  runInTrackingTransaction = <T>(fn: () => T, debugLabel?: string | false) => {
     beginTrackingTransaction!(debugLabel);
+    let didError = true;
 
     try {
-      fn();
+      let value = fn();
+      didError = false;
+      return value;
     } finally {
-      endTrackingTransaction!();
+      if (didError !== true) {
+        endTrackingTransaction!();
+      }
```

Comment:
This seems a tad surprising, since `runInTrackingTransaction` always calls `beginTrackingTransaction` why wouldn't it always call `endTrackingTransaction`?

Is it because we already call `resetTracking` (basically in our callsite)?

Reply:
Exactly that. With the current structure of things, there is a case where this can happen unfortunately. I think we can work on restructuring it in the future to make it a bit more bulletproof.